### PR TITLE
Admin Generator (Future): Add support for filterOperators in the grid

### PR DESCRIPTION
**Pull request by Ben-Ho**
## Requirement
Allow using filterOperators for generated grids. This is for example required if custom filters are needed or you want to add some specific filter to a column. Generated code conforms with https://github.com/vivid-planet/comet/pull/2443.

## Solution
Add `filterOperators`-field with ImportReference-type to provide custom-filter component.
For the filter added it was also required to add an option (`fieldName`) to overwrite what's set to mui-data-grid `field`-prop, because it's currently not possible to use `manufacturer` as name. `manufacturer` is not listed in `UsableFields`-type because it's an object. But the filter-name is `manufacturer` because it's a relation-filter.

## Open issues
- [x] [columnField](https://github.com/vivid-planet/comet/pull/2446/files#diff-644829f485d69e1788b7291bde71a325aabc663862daa095d805e2e471a44a8cR58) needs to match `name` or `fieldName` for the filter to work correctly. This could probably be a prop, but probably nobody would do this for a specific filter in handmade.

### CODE REVIEW

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts [hunk 1]
```diff
@@ -248,6 +248,21 @@ export function generateGrid(
 
         let gridType: "number" | "boolean" | "dateTime" | "date" | undefined;
 
+        let filterOperators: string | undefined;
+        if (column.filterOperators) {
+            let importPath = column.filterOperators.import;
+            if (importPath.startsWith("./")) {
+                //go one level up as generated files are in generated subfolder
+                importPath = `.${importPath}`;
+            }
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/generator/future/generateGrid.ts:257]: Do we perform this trick for all import statements?

**Comment by johnnyomair** [comment]
> manufacturer is not listed in UsableFields-type because it's an object.

Could/should this be changed?

**Comment by johnnyomair** [comment]
> [columnField](https://github.com/vivid-planet/comet/pull/2446/files#diff-644829f485d69e1788b7291bde71a325aabc663862daa095d805e2e471a44a8cR58) needs to match name or fieldName for the filter to work correctly. This could probably be a prop, but probably nobody would do this for a specific filter in handmade.

I don't understand this (it seems to me that the link is wrong). Could you please explain it with more detail?

**Comment by Ben-Ho** [comment]
> > manufacturer is not listed in UsableFields-type because it's an object.
> 
> Could/should this be changed?

this was intentionally changed, because it's no dedicated field to render. And the type I prior came up with had issues with typescript warnings if used with any. `GridConfig<any>`

@nsams what do you think?

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts (hunk 1 above)
Comment (Ben-Ho) [reply packages/admin/cms-admin/src/generator/future/generateGrid.ts:257]: as far as I know this is only used for ImportReference-type in config, like it's used in form-generator to generate a reference to field validation-code. probably also for generated graphql-query-files, but I would need to check this... (I copied this from [here](https://github.com/vivid-planet/comet/blob/main/packages/admin/cms-admin/src/generator/future/generateForm/generateFormField.ts#L62))

**Comment by Ben-Ho** [comment]
> > [columnField](https://github.com/vivid-planet/comet/pull/2446/files#diff-644829f485d69e1788b7291bde71a325aabc663862daa095d805e2e471a44a8cR58) needs to match name or fieldName for the filter to work correctly. This could probably be a prop, but probably nobody would do this for a specific filter in handmade.
> 
> I don't understand this (it seems to me that the link is wrong). Could you please explain it with more detail?

ah, yes, that's the wrong location. it's about `applyValue` call in ManufacturerFilter.tsx -> `onChange`. `columnField` needs to match the `field` string of the mui-data-grid-column-configuration the filterOperator is added to, for mui-data-grid to correctly initialize filterOperator with value/filter-formState from url (learned and checked that by trial and error). _(Note: I don't like code loosely depending on some other code, because how am I supposed to guess the connection if altering or copying and modifying this code to implement a new custom-filter.)_

maybe you got an idea how to connect those two locations in any way?

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts (hunk 1 above)
Comment (johnnyomair) [reply packages/admin/cms-admin/src/generator/future/generateGrid.ts:257]: I see... then it's okay 🙂

**Comment by johnnyomair** [comment]
It's odd that you need to specify the column name in the applyFilter callback. They're not doing this in the docs: https://mui.com/x/react-data-grid/filtering/customization/#customize-the-operators. Are you sure this is required?

**Comment by Ben-Ho** [comment]
> It's odd that you need to specify the column name in the applyFilter callback. They're not doing this in the docs: https://mui.com/x/react-data-grid/filtering/customization/#customize-the-operators. Are you sure this is required?

yes I'm sure: https://v6.mui.com/x/api/data-grid/grid-filter-item/

in v7 it's called differently: https://mui.com/x/api/data-grid/grid-filter-item/#grid-filter-item-prop-field

**Comment by johnnyomair** [comment]
In the MUI example they use a object spread. Could we do this instead?

```ts
const handleFilterChange: RatingProps['onChange'] = (event, newValue) => {
    applyValue({ ...item, value: newValue });
  };
```

https://v6.mui.com/x/react-data-grid/filtering/customization/#create-a-custom-operator

**Review by johnnyomair — APPROVED** [review]

**Review by johnnyomair — APPROVED** [review]

**Comment by Ben-Ho** [comment]
@johnnyomair updated ManufacturerFilter.tsx to match demo-pull-request

File: packages/admin/cms-admin/src/generator/future/generateGrid.ts [hunk 2]
```diff
@@ -324,6 +324,21 @@ export function generateGrid(
 
         let gridType: "number" | "boolean" | "dateTime" | "date" | undefined;
 
+        let filterOperators: string | undefined;
+        if (column.type !== "combination" && column.filterOperators) {
```
Comment (nsams) [inline packages/admin/cms-admin/src/generator/future/generateGrid.ts:328]: why type != "combination"?

if an explicit filterOperators is specified, shouldn't we be able to use it?

**Review by johnnyomair — APPROVED** [review]

**Review by nsams — APPROVED** [review]

**Pull request merged**